src/org/zaproxy/zap/extension/allinonenotes/ExtensionAllInOneNotes.java

```diff
@@ -0,0 +1,230 @@
+            reload.addActionListener(l -> {
+                hookView.addStatusPanel(getStatusPanel());
+            });
+
+        }
+    }
+
+    @Override
+    public boolean canUnload() {
+        // The extension can be dynamically unloaded, all resources used/added can be freed/removed from core.
+        return true;
+    }
+
+    @Override
+    public void unload() {
+        super.unload();
+        // Unloading the event consumer
+        ZAP.getEventBus().unregisterConsumer(eventConsumerImpl, HistoryReferenceEventPublisher.getPublisher().getPublisherName());
+    }
+
+    @Override
+    public void sessionChanged(final Session session)  {
+        // session changed - extension should basically act as if "Reload" has been called and re-build the status panel
+        hookView.addStatusPanel(getStatusPanel());
+    }
+
+    @Override
+    public void sessionAboutToChange(Session session) {
+        //Left empty for now
+    }
+
+    @Override
+    public void sessionScopeChanged(Session session) {
+        //Left empty for now
+    }
+
+    @Override
+    public void sessionModeChanged(Control.Mode mode) {
+        //Left empty for now
+    }
+
+    private AbstractPanel getStatusPanel() {
+
+        NotesTableModel notesTableModel = new NotesTableModel();
+        notesTable.setModel(notesTableModel);
+
+        ExtensionHistory extHist = (ExtensionHistory) org.parosproxy.paros.control.Control.getSingleton().
+                getExtensionLoader().getExtension(ExtensionHistory.NAME);
+
+
+        if (extHist != null) {
+
+            int LastHistoryId = extHist.getLastHistoryId();
+
+            int i=0;
+            while (i++ < LastHistoryId) {
+                HistoryReference hr = extHist.getHistoryReference(i);
+                if (hr != null) {
+                    if (hr.hasNote()) {
+                        eventConsumerImpl.addRowToNotesTable(i);
+                    }
+                }
+            }
+        }
+
+        notesTable.setColumnSelectionAllowed(false);
+        notesTable.setCellSelectionEnabled(false);
+        notesTable.setRowSelectionAllowed(true);
+        notesTable.setAutoCreateRowSorter(true);
+        notesTable.setColumnControlVisible(true);
+        notesTable.getSelectionModel().addListSelectionListener(new NotesTableSelectionHandler(notesTable, extHist));
+
+        exportButton = new TableExportButton<>(notesTable);
+
+        JPanel buttonContainer = new JPanel();
+        buttonContainer.setLayout(new BoxLayout(buttonContainer, BoxLayout.X_AXIS));
+        buttonContainer.add(reload);
+        buttonContainer.add(exportButton);
+
+        JPanel pContainer = new JPanel();
+        pContainer.setLayout(new BoxLayout(pContainer, BoxLayout.Y_AXIS));
+        pContainer.add(buttonContainer);
+        pContainer.add(new JScrollPane(notesTable));
+
+        if (statusPanel != null) {
+            statusPanel.removeAll();
```

Comment:
Would it make sense to move the history reference loop from above to an update/refresh/reload function that just refreshes the table. Make statusPanel into a member of this class that's setup during init. Instead of removing and re-initializing everything on each reload? (Note: I haven't tested this PR branch, or played with the code at all....this is just me reading code in a web browser.)